coderd/database/migrations/000034_linked_user_id.up.sql

```diff
@@ -0,0 +1,20 @@
+BEGIN;
+
+ALTER TABLE users ADD COLUMN login_type login_type NOT NULL DEFAULT 'password'; 
+ALTER TABLE users ADD COLUMN linked_id text NOT NULL DEFAULT ''; 
+
+UPDATE 
+  users
+SET
+  login_type = (
+    SELECT 
+      login_type
+    FROM
+      api_keys
+    WHERE
+      api_keys.user_id = users.id
+    ORDER BY updated_at DESC
+    LIMIT 1
+  );
```

Comment:
I was thinking we might want this as an `user_links` table instead.

In the future, it seems reasonable that developers would sign in with Okta, but authenticate with GitHub or something for us to inject an SSH key.

I think we also want to store the secret and access key for OIDC refresh in this table instead of attached to an API key. Having numerous refreshes that could occur (if the user had like 100 API keys) seems like unintended behavior.

What do you think?

Reply:
Yeah I think that makes sense. I think being able to log in with multiple types of providers is a little weird. What if you have a single user with two different emails for Github and Okta? Are we just going to have a rule that you have to have a matching email in order to login as the same user?

I think it's also worth noting that with this PR the `AllowSignups` flag for both Github and ODIC is a bit obviated. Since you can't create a `built-in` user and then have them sign in with OIDC. I don't see that as a big deal since v1 works that way and we haven't had any complaints.

Reply:
I feel like we could define that mapping in a pretty abstract way. The links table is less about multiple to me, and more about associating a link with a specific set of access tokens representing the connection.

We don't need to abstract the mapping right now. We can just leave it as-is with the single provider and somehow do the mapping in the future. I don't think we'd do email matching... more-so just allow linking alternative accounts when already authenticated.

Reply:
gotcha so move all the oauth related columns to this new table...should we call it `oauth_links`?

Reply:
Let's keep it more generalized for now like `user_links`.